internal/provider/resource_offer.go

```diff
@@ -2,173 +2,255 @@ package provider
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	frameworkdiags "github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	frameworkResSchema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
 	"github.com/juju/terraform-provider-juju/internal/juju"
 )
 
-func resourceOffer() *schema.Resource {
-	return &schema.Resource{
-		// This description is used by the documentation generator and the language server.
-		Description: "A resource that represent a Juju Offer.",
+// Ensure provider defined types fully satisfy framework interfaces.
+var _ resource.Resource = &offerResource{}
+var _ resource.ResourceWithConfigure = &offerResource{}
+var _ resource.ResourceWithImportState = &offerResource{}
 
-		CreateContext: resourceOfferCreate,
-		ReadContext:   resourceOfferRead,
-		DeleteContext: resourceOfferDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: schema.ImportStatePassthroughContext,
-		},
-		Schema: map[string]*schema.Schema{
-			"model": {
+func NewOfferResource() resource.Resource {
+	return &offerResource{}
+}
+
+type offerResource struct {
+	client *juju.Client
+}
+
+type offerResourceModel struct {
+	ModelName       types.String `tfsdk:"model"`
+	OfferName       types.String `tfsdk:"name"`
+	ApplicationName types.String `tfsdk:"application_name"`
+	EndpointName    types.String `tfsdk:"endpoint"`
+	URL             types.String `tfsdk:"url"`
+	// ID required by the testing framework
+	ID types.String `tfsdk:"id"`
+}
+
+func (o *offerResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_offer"
+}
+
+func (o *offerResource) Schema(_ context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = frameworkResSchema.Schema{
+		Description: "A resource that represent a Juju Offer.",
+		Attributes: map[string]frameworkResSchema.Attribute{
+			"model": frameworkResSchema.StringAttribute{
 				Description: "The name of the model to operate in.",
-				Type:        schema.TypeString,
 				Required:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
 			},
-			"name": {
+			"name": frameworkResSchema.StringAttribute{
 				Description: "The name of the offer.",
-				Type:        schema.TypeString,
 				Optional:    true,
 				Computed:    true,
-				ForceNew:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
```

Comment:
RequiresReplaceIfConfigured() and stringplanmodifier.UseStateForUnknown(),

Reply:
Did the extra `stringplanmodifier.RequiresReplaceIfConfigured(),` at line 57 belong here instead?